### DISCUSSION TITLE
Trigger switch on time elapsed in ROI using continuous bool stream

### DISCUSSION THREAD

**Discussion by ChucklesOnGitHub** [bonsai-rx/bonsai#1278] [Q&A] [answered]
Hi!

I am working on a game in which sound is generated based on the position of the arms using DLC-live. This is working, but I want to add the ability to stop and start the sound based on a condition. If a wrist spends 2 seconds in a ROI, the sound starts and can be controlled by the arms (game is on). Next time the wrist spends 2 seconds in the ROI, the sound amplitude goes to zero (game is muted), but the pose is still estimated and the workflow keeps running.

I cannot work out the toggle condition.
Since the sound amplitude value has to be continuously passed to the function generator, I think I have to have a continuous reporter stream of bools saying whether the game should be on or muted. 
-the toggle condition is (inROI) & (timeElapsed)
-when toggle condition is met, the reporter stream changes to opposite of what it was (on/muted).

To make it an easier workflow, I’m testing it with mousemove. The (inROI) condition is a threshold on one dimension.
a) **How to test for the condition timeElapsed?** I used Accumulate to count frames inROI but I couldn't restart the count each time the inROI condition was met. I tried Mod but this of course gives a periodic toggle and doesn't work.
b) **How to generate the “reporter stream”?** Do I even need it?

It was much easier without the timeElapsed condition, since just using GreaterThan gave me a continuous bool stream. I tried a bunch of trigger solutions from the forum but they all worked non-continuously.

There's some part of the logic I can't wrap my head around, thanks for any pointers!

[https://github.com/ChucklesOnGitHub/dlc-fun/blob/main/theremin/Theremin_amplitude_toggle_debug.bonsai](url)

**Comment by bruno-f-cruz** [comment]
Sounds like a cool project! I would love to take a look but unfortunately, your link for the workflow does not work ...
Also, I have recently pushed a rc version of an operator that implements a few [MoveNet](https://www.tensorflow.org/hub/tutorials/movenet) networks for human tracking. Ping me on discord if you want to give it a try! 
B

**Comment by ChucklesOnGitHub** [comment]
[Here](https://github.com/ChucklesOnGitHub/dlc-fun/blob/main/theremin/Theremin_amplitude_toggle_debug.bonsai) it is again, I had messed up the url syntax. Thank you, I'll catch you on Discord!

**Comment by bruno-f-cruz** [comment, marked as answer]
I refactored the workflow a bit and added the logic you described (I think...)
[Refactored1278.zip](https://github.com/bonsai-rx/bonsai/files/10943718/Refactored1278.zip)
![image](https://user-images.githubusercontent.com/7049351/224363861-80452374-6ce8-4874-9e44-818cdb7a8bba.png)
Briefly:
We get the MouseData and convert it to a `Point`. We then apply a `ContainsPoint` operator on the Tuple created from the point and a region of interest. It will return `True` if the point is inside and false otherwise. On each toggle (`DistinctUntilChanged`) we start an `Observable` with a `Timer` inside. Outside this timer, we have a `Switch` that basically implements something akin to: "Each time a toggle occurs, I will discard my current timer and start a new one". This effectively makes the sequence only produce values if the current value has not changed for N seconds. After that, we filter rising-edges (`Condition`) (when the computer mouse enters the ROI) and toggle the game state value `IsGameActive`.
We then use this game state to toggle your previous logic.

Not sure if it will work exactly the way you had in mind, but hope it helps :)
B

**Reply by ChucklesOnGitHub** [reply]
Thank you, this is brilliant, it worked as I expected. I only changed the ifFalse so it is completely silent when the game is off.

I very close to implementing the same logic with the pose estimation (now using MoveNet), but am having a type conversion error I just can't figure out: I seem to have System.Single and System.Tuple`2[System.Int32,System.Int32] for the same published subject, but I thought I was indexing for a float inside the tuple. Would you mind helping me out with that too please?

[Theremin_movenet_withcontrollogic.zip](https://github.com/bonsai-rx/bonsai/files/10948914/Theremin_movenet_withcontrollogic.zip)

Edit: I think the problem might be that I don't know how to use multicast

**Reply by ChucklesOnGitHub** [reply]
That was it, reading https://bonsai-rx.org/docs/articles/subjects.html helped :')
[Theremin_movenet_withcontrollogic_multicast.zip](https://github.com/bonsai-rx/bonsai/files/10949086/Theremin_movenet_withcontrollogic_multicast.zip)
